TSS-3148: Create the BaseScript class

r-owen opened the pull request.

timj commented on setup.cfg:
```diff
@@ -0,0 +1,8 @@
+[flake8]
+max-line-length = 110
+ignore = E133, E226, E228, N802, N803, N806
```
Does telescope and site have the same ignores as DM? In theory T&S could be different.

r-owen replied:
T&S has no standard so I adopted DM standards.

timj commented on tests/test_script_loader.py:
```diff
@@ -22,40 +22,54 @@
-            info_coro = self.remote.evt_script_info.next(script_info_data)
-            id_ack = await self.remote.cmd_load.start(load_data, timeout=5)
+            load_data.config = "wait_time: 1"
+            info_coro = self.remote.evt_script_info.next(timeout=2)
```
`next` as a method name could be a bit confusing because of the `next()` Python built-in function.

r-owen replied:
I agree, but I think the name is clear enough that I'm inclined to keep it. If you disagree strongly feel free to file a ticket against salobj

timj commented on python/scriptloader/script_loader.py:
```diff
@@ -60,7 +60,7 @@ class ScriptLoader(salobj.BaseCsc):
       configuration specified in the ``load`` command.
     * Configuring the script causes it to output an estimate of its
-      duration and puts the script into the CONFIGURED state,
+      duration and puts the script into the `ScriptState.CONFIGURED` state,
```
Only duration? Or all the information that the queue might need?

r-owen replied:
Good catch. Left over from metadata implementation. I'll update it.

tribeiro commented:
How one could run the script loader CSC? I think it would be a good idea to add a `scripts/script_loader.py` that will start the script loader. Unless, it already existing but I couldn't find it.

r-owen commented on doc/scriptloader/index.rst:
```diff
@@ -0,0 +1,41 @@
+* If the ``configure`` command succeeds, the script must report the ``metadata`` event and go into the ``CONFIGURED`` state.
+* The ``configure`` command is permitted in both the ``UNCONFIGURED`` and the ``CONFIGURED`` states.
+* The ``run`` command runs the script. The script must be in the ``CONFIGURED`` state to be run.
+* When the ``run`` command finishes the script must exit, after reporting one of three states: ``DONE``, ``STOPPED`` or ``FAILED``, as appropriate.
```
Yes

r-owen commented on doc/conf.py:
```diff
@@ -4,10 +4,10 @@
 
 from documenteer.sphinxconfig.stackconf import build_package_configs
-import lsst.scriptrunner.version
+import scriptloader.version
```
Good question. I've been avoiding it because T&S doesn't use any of DM, including the bit that allows shared sub-namespaces. But I think we should revisit that. I would also prefer to see DM code and T&S code in separate sub namespaces, since DM has so many packages.

r-owen replied:
Also I think T&S mostly uses ts_ name prefix for this reason. I should start using that, and will use it for the queue package.

r-owen commented on doc/scriptloader/index.rst:
```diff
@@ -0,0 +1,41 @@
+Scripts
+=======
+
+A script is a SAL component that supports the Script API. In addition:
```
Script is a SAL component name. I think SALScript would be redundant there. But I suppose SAL script could be the generic name. Would that suffice?

r-owen commented on python/scriptloader/script_loader.py:
```diff
@@ -37,9 +37,9 @@ class ScriptLoader(salobj.BaseCsc):
     ----------
     standardpath : `str`, `bytes` or `os.PathLike`
-        Path to standard modules.
+        Path to standard scripts.
```
You mean like "SAL script"?

r-owen commented on python/scriptloader/test_utils.py:
```diff
@@ -0,0 +1,74 @@
+        if self.wait_time < 0:
+            raise salobj.ExpectedError(f"wait_time={self.wait_time} must be >= 0")
+        self.fail_run = bool(fail_run)
+        self.fail_cleanup = bool(fail_cleanup)
```
My thought was: any Python object can be evaluated as a boolean, but I would rather have an actual True or False for the value. Do you have an alternate recommendation?

r-owen commented:
@tribeiro good point. I added bin.src/run_script_loader.py which scons rewrites as bin/run_script_loader.py and added bin to the PATH in the ups table file.

r-owen commented on doc/scriptloader/index.rst:
```diff
@@ -0,0 +1,41 @@
+
+A script is a SAL component that supports the Script API. In addition:
+
+* A script is a command-line executable that takes a single command line argument: the index of the SAL component.
```
I agree. I think there is no safe way to protect against badly written scripts that start running when they should not, or talk to components they should not. We could have a bit more protection if we insisted all scripts are Python and have a standard name, as we could load a specified module and look for the standard name. But it's not much protection.
